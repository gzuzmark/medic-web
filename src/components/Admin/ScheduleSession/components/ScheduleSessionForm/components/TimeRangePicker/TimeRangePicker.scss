.TimeRangePicker {
  &:last-child {
    .TimeRangePicker_option--add {
      display: flex;
    }
  }
  &_options {
    display: flex;
    justify-content: space-between;
    margin-top: 28px;
  }
  &_option {
    align-items: center;
    cursor: pointer;
    display: none;
    padding: 0 8px;
    &--delete {
      display: flex;
    }
    &:hover {
      span {
        color: #773ad1!important;
      }
      .icon {
        fill: #773ad1!important;
        border-color: #773ad1!important;
      }

    }
  }
  &_option-text {
    margin-left: 9px;
  }
}