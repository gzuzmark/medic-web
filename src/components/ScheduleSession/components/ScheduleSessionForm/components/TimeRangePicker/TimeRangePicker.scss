.TimeRangePicker {
  &:first-child:nth-last-child(1) {
    .TimeRangePicker_option--delete {
      display: none;
    }
  }
  &:last-child {
    .TimeRangePicker_option--add {
      display: flex;
    }
  }
  &_options {
    display: flex;
    justify-content: space-between;
    margin-top: 28px;
  }
  &_option {
    align-items: center;
    cursor: pointer;
    display: none;
    padding: 0 8px;
    &--delete {
      display: flex;
    }
  }
  &_option-text {
    margin-left: 9px;
  }
}