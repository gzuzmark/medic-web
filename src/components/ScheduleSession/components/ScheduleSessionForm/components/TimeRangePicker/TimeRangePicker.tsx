import * as moment from 'moment';
import * as React from 'react';
import TimePicker from "../../../../../../common/TimePicker/TimePicker";
import InputDatePicker from "../../../../../Reports/components/InputDatePicker/InputDatePicker";
import FormColumn from "../../../FormRow/components/FormColumn/FormColumn";
import FormRow from "../../../FormRow/FormRow";
// import './TimeRangePicker.scss';

interface IPropsTimeRangePicker {
    date: Date;
    onChange: (params: object) => void;
}

const formatTime = (date: moment.Moment) => {
    const format = 15;
    const minutes = date.get('minutes');
    return minutes + (format  - minutes % format)
};

const getInitDate = (date: Date, limit: Date) => {
    const initDate = moment(date);
    const endDate = moment(date);
    const limitDate = moment(limit);
    if (initDate.format("YYYY-MM-DD") === limitDate.format("YYYY-MM-DD")) {
        const extraMinutes = formatTime(initDate);
        initDate.set('minutes', extraMinutes);
    } else {
        initDate.set('minutes', 0);
        initDate.set('hours', 0);
    }
    initDate.set('seconds', 0);
    endDate.set('hours', 23);
    endDate.set('minutes', 59);
    endDate.set('seconds', 0);
    return { endDate, initDate }
};

const TimeRangePicker: React.StatelessComponent<IPropsTimeRangePicker> = (props) => {
    let counter = 0;
    const initDates = getInitDate(props.date, new Date());
    const endDates = getInitDate(props.date, new Date());
    return (
        <div style={{marginTop: 30}}>
            <FormRow columns={[
                <FormColumn key={`TimeRangePicker_${++counter}`}  width={4}>
                    <InputDatePicker
                        id={'startDate'}
                        date={props.date}
                        updateState={props.onChange} />
                </FormColumn>,
                <FormColumn key={`TimeRangePicker_${++counter}`} width={5}>
                    <TimePicker
                        defaultText={"11:00 am"}
                        name={'Inicia'}
                        onChange={props.onChange}
                        from={initDates.initDate}
                        to={initDates.endDate}/>
                </FormColumn>,
                <FormColumn key={`TimeRangePicker_${++counter}`} width={5}>
                    <TimePicker
                        defaultText={"11:00 am"}
                        name={'Termina'}
                        onChange={props.onChange}
                        from={endDates.initDate}
                        to={endDates.endDate}/>
                </FormColumn>,
                <FormColumn key={`TimeRangePicker_${++counter}`} width={3}>
                    &nbsp;
                </FormColumn>,
            ]} />
        </div>

    );
};

export default TimeRangePicker;


