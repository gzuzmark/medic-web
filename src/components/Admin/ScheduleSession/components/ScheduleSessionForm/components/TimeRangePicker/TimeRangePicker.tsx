import * as moment from 'moment';
import * as React from 'react';
import ConsoleColor from "../../../../../../../common/ConsoleColor";
import { Text } from '../../../../../../../common/ConsoleText';
import FormColumn from "../../../../../../../common/FormRow/components/FormColumn/FormColumn";
import FormRow from "../../../../../../../common/FormRow/FormRow";
import Icon from "../../../../../../../common/Icon/Icon";
import TimePicker from "../../../../../../../common/TimePicker/TimePicker";
import {IListItem} from "../../../../../../../domain/Lists";
import InputDatePicker from "../../../../../Reports/components/InputDatePicker/InputDatePicker";
import './TimeRangePicker.scss';

interface IPropsTimeRangePicker {
    id: number;
    total: number;
    onChange: (id: number, from: Date | null, to: Date | null, key: string) => void;
    onRemoveWorkshop: (id: number) => void;
    onAddWorkshop: (from: Date | null, to: Date | null) => void;
    uniqueKey: string;
}

interface IFormatDates {
    endDate: moment.Moment;
    initDate: moment.Moment;
}

interface IStateTimeRangePicker {
    currentDate: Date;
    fromDates: IFormatDates;
    toDates: IFormatDates;
    selectedFromDate: Date | null;
    selectedToDate: Date | null;
}


class TimeRangePicker extends React.Component<IPropsTimeRangePicker, {}> {
    public state: IStateTimeRangePicker;
    private actionColor: string;
    constructor(props: IPropsTimeRangePicker) {
        super(props);
        const currentDate = new Date();
        this.state = {
            currentDate,
            fromDates: this.getInitDate(currentDate),
            selectedFromDate: null,
            selectedToDate: null,
            toDates: this.getInitDate(currentDate),
        };
        this.onChangeDateBase = this.onChangeDateBase.bind(this);
        this.onSelectFromDate = this.onSelectFromDate.bind(this);
        this.onSelectToDate = this.onSelectToDate.bind(this);
        this.onAdd = this.onAdd.bind(this);
        this.onRemove = this.onRemove.bind(this);
        this.actionColor = ConsoleColor.TEXT_COLORS.actionColor;
    }

    public render() {
        let counter = 0;
        const initDateName = this.state.selectedFromDate ?
            moment(this.state.selectedFromDate).format('hh:mm a') : '';
        const endDateName = this.state.selectedToDate ?
            moment(this.state.selectedToDate).format('hh:mm a') : '';
        return (
            <div className={'TimeRangePicker'} style={{marginTop: 30}}>
                <FormRow columns={[
                    <FormColumn key={`TimeRangePicker_${++counter}`}  width={4}>
                        <Text className='FormSession-label'>Fecha</Text>
                        <InputDatePicker
                            id={'startDate'}
                            date={this.state.currentDate}
                            updateState={this.onChangeDateBase} />
                    </FormColumn>,
                    <FormColumn key={`TimeRangePicker_${++counter}`} width={5}>
                        <Text className='FormSession-label'>
                            <Icon name={'clock'}/> Inicia</Text>
                        <TimePicker
                            defaultText={"11:00 am"}
                            name={initDateName}
                            onChange={this.onSelectFromDate}
                            from={this.state.fromDates.initDate.toDate()}
                            to={this.state.fromDates.endDate.toDate()}/>
                    </FormColumn>,
                    <FormColumn key={`TimeRangePicker_${++counter}`} width={5}>
                        <Text className='FormSession-label'>
                            <Icon name={'clock'}/> Termina</Text>
                        <TimePicker
                            defaultText={"12:00 pm"}
                            name={endDateName}
                            onChange={this.onSelectToDate}
                            from={this.state.toDates.initDate.toDate()}
                            to={this.state.toDates.endDate.toDate()}/>
                    </FormColumn>,
                    <FormColumn key={`TimeRangePicker_${++counter}`} width={3}>
                        <div className='TimeRangePicker_options'>
                            {this.props.id !== 4 &&
                            <div className='TimeRangePicker_option TimeRangePicker_option--add' onClick={this.onAdd}>
                                <Icon name='add'
                                      style={{fill: this.actionColor, borderRadius: '50%', border: `1px solid ${this.actionColor}`}}/>
                                <Text color='actionColor' className='TimeRangePicker_option-text'>Agregar horario</Text>
                            </div>}
                            {this.props.total > 1 &&
                            <div className='TimeRangePicker_option TimeRangePicker_option--delete' onClick={this.onRemove}>
                                <Icon name='trash' style={{fill: this.actionColor}} />
                                <Text color='actionColor' className='TimeRangePicker_option-text'>Eliminar</Text>
                            </div>}
                        </div>
                    </FormColumn>,
                ]} />
            </div>
        )
    }

    private onRemove() {
        this.props.onRemoveWorkshop(this.props.id);
    }

    private onAdd() {
        this.props.onAddWorkshop(null, null);
    }

    private updateWorkshops() {
        this.props.onChange(this.props.id, this.state.selectedFromDate, this.state.selectedToDate, this.props.uniqueKey);
    }

    private updateDate(date: Date, key: string) {
        this.setState({[key]: date}, () => {
            this.updateWorkshops();
        });
    }

    private onSelectToDate(item: IListItem) {
        this.updateDate(new Date(item.id), 'selectedToDate');
    }

    private onSelectFromDate(item: IListItem) {
        const fromDate = new Date(item.id);
        const toDate = new Date(item.id);
        toDate.setMinutes(toDate.getMinutes() + 45);
        const toDates = this.getInitDate(toDate, true);
        this.setState({
            selectedToDate: null,
            toDates
        }, () => {
            this.updateDate(fromDate, 'selectedFromDate');
        });
    }

    private onChangeDateBase(newDate: {startDate: Date}) {
        const currentDate = newDate.startDate;
        const initDates = this.getInitDate(currentDate);
        this.setState({
            currentDate,
            fromDates: initDates,
            selectedFromDate: null,
            selectedToDate: null,
            toDates: initDates
        }, () => {
            this.updateWorkshops();
        });
    }

    private getInitDate(date: Date, restrict = false) {
        let initDate = moment(date);
        const endDate = moment(date);
        const limitDate = moment();
        const isToday = initDate.format("YYYY-MM-DD") === limitDate.format("YYYY-MM-DD");
        let initMinutes = 0;
        let endMinutes= 0;
        if (restrict) {
            initMinutes = initDate.get('minutes');
            endMinutes = 59;
        } else if (isToday) {
            initDate = moment();
            initMinutes = this.formatTime(limitDate);
        } else {
            initDate.set('hours', 8);
        }
        initDate.set('minutes', initMinutes).set('seconds', 0);
        endDate.set('hours', 23).set('minutes', endMinutes).set('seconds', 0);
        return { endDate, initDate }
    }

    private formatTime(date: moment.Moment): number {
        const format = 15;
        const minutes = date.get('minutes');
        return minutes + (format  - minutes % format)
    }


}

export default TimeRangePicker;


