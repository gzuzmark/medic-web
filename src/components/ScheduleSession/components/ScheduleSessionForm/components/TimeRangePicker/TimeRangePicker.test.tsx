import { shallow } from 'enzyme';
import 'jest-localstorage-mock';
import * as React from 'react';
import TimeRangePicker from "./TimeRangePicker";


describe.skip('TimeRangePicker Test',() => {
    let props: any;
    let mountedTimeRangePicker: any;
    const getComponent = () => {
        if (!mountedTimeRangePicker) {
            mountedTimeRangePicker = shallow(
                <TimeRangePicker {...props} />
            );
        }
        return mountedTimeRangePicker;
    };

    beforeEach(() => {
        props = {
            date: new Date(Date.UTC(2012,3,13,1,31,38)),
            onChange: void(0)
        };
        mountedTimeRangePicker = undefined;
    });

    it("render: render TimeRangePicker default", () => {
        const component = getComponent();
        expect(component).toMatchSnapshot();
    });

});