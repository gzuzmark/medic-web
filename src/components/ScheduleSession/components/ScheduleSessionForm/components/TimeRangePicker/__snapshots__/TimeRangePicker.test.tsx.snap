// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeRangePicker Test render: render TimeRangePicker default 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TimeRangePicker
    date={2012-04-13T01:31:38.000Z}
    onChange={undefined}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <FormRow
        columns={
          Array [
            <FormColumn
              width={4}
            >
              <InputDatePicker
                date={2012-04-13T01:31:38.000Z}
                id="startDate"
                updateState={undefined}
              />
            </FormColumn>,
            <FormColumn
              width={5}
            >
              <TimePicker
                defaultText="11:00 am"
                from={"2012-04-12T05:00:00.000Z"}
                name="Inicia"
                onChange={undefined}
                step={15}
                to={"2012-04-13T04:59:00.000Z"}
              />
            </FormColumn>,
            <FormColumn
              width={5}
            >
              <TimePicker
                defaultText="11:00 am"
                from={"2012-04-12T05:00:00.000Z"}
                name="Termina"
                onChange={undefined}
                step={15}
                to={"2012-04-13T04:59:00.000Z"}
              />
            </FormColumn>,
            <FormColumn
              width={3}
            >
               
            </FormColumn>,
          ]
        }
      />,
      "style": Object {
        "marginTop": 30,
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "columns": Array [
          <FormColumn
            width={4}
          >
            <InputDatePicker
              date={2012-04-13T01:31:38.000Z}
              id="startDate"
              updateState={undefined}
            />
          </FormColumn>,
          <FormColumn
            width={5}
          >
            <TimePicker
              defaultText="11:00 am"
              from={"2012-04-12T05:00:00.000Z"}
              name="Inicia"
              onChange={undefined}
              step={15}
              to={"2012-04-13T04:59:00.000Z"}
            />
          </FormColumn>,
          <FormColumn
            width={5}
          >
            <TimePicker
              defaultText="11:00 am"
              from={"2012-04-12T05:00:00.000Z"}
              name="Termina"
              onChange={undefined}
              step={15}
              to={"2012-04-13T04:59:00.000Z"}
            />
          </FormColumn>,
          <FormColumn
            width={3}
          >
             
          </FormColumn>,
        ],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <FormRow
          columns={
            Array [
              <FormColumn
                width={4}
              >
                <InputDatePicker
                  date={2012-04-13T01:31:38.000Z}
                  id="startDate"
                  updateState={undefined}
                />
              </FormColumn>,
              <FormColumn
                width={5}
              >
                <TimePicker
                  defaultText="11:00 am"
                  from={"2012-04-12T05:00:00.000Z"}
                  name="Inicia"
                  onChange={undefined}
                  step={15}
                  to={"2012-04-13T04:59:00.000Z"}
                />
              </FormColumn>,
              <FormColumn
                width={5}
              >
                <TimePicker
                  defaultText="11:00 am"
                  from={"2012-04-12T05:00:00.000Z"}
                  name="Termina"
                  onChange={undefined}
                  step={15}
                  to={"2012-04-13T04:59:00.000Z"}
                />
              </FormColumn>,
              <FormColumn
                width={3}
              >
                 
              </FormColumn>,
            ]
          }
        />,
        "style": Object {
          "marginTop": 30,
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "columns": Array [
            <FormColumn
              width={4}
            >
              <InputDatePicker
                date={2012-04-13T01:31:38.000Z}
                id="startDate"
                updateState={undefined}
              />
            </FormColumn>,
            <FormColumn
              width={5}
            >
              <TimePicker
                defaultText="11:00 am"
                from={"2012-04-12T05:00:00.000Z"}
                name="Inicia"
                onChange={undefined}
                step={15}
                to={"2012-04-13T04:59:00.000Z"}
              />
            </FormColumn>,
            <FormColumn
              width={5}
            >
              <TimePicker
                defaultText="11:00 am"
                from={"2012-04-12T05:00:00.000Z"}
                name="Termina"
                onChange={undefined}
                step={15}
                to={"2012-04-13T04:59:00.000Z"}
              />
            </FormColumn>,
            <FormColumn
              width={3}
            >
               
            </FormColumn>,
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
